File: zetasql/public/types/type_parameters_test.cc
//
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "zetasql/public/types/type_parameters.h"

#include "google/protobuf/descriptor.h"
#include "zetasql/public/type.pb.h"
#include "zetasql/public/type_parameters.pb.h"
#include "zetasql/public/types/type_factory.h"
#include "zetasql/testdata/test_schema.pb.h"
#include "gmock/gmock.h"
#include "gtest/gtest.h"
#include "absl/strings/string_view.h"

namespace zetasql {
namespace {

TEST(TypeParameters, CreateStringTypeParametersWithMaxLiteral) {
  StringTypeParametersProto proto;
  proto.set_is_max_length(true);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters type_param,
                       TypeParameters::MakeStringTypeParameters(proto));
  EXPECT_TRUE(type_param.IsStringTypeParameters());
  EXPECT_EQ(type_param.string_type_parameters().is_max_length(), true);
  EXPECT_EQ(type_param.DebugString(), "(max_length=MAX)");
}
TEST(TypeParameters, CreateStringTypeParametersWithMaxLength) {
  StringTypeParametersProto proto;
  proto.set_max_length(1000);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters type_param,
                       TypeParameters::MakeStringTypeParameters(proto));
  EXPECT_TRUE(type_param.IsStringTypeParameters());
  EXPECT_EQ(type_param.string_type_parameters().max_length(), 1000);
  EXPECT_EQ(type_param.DebugString(), "(max_length=1000)");
}
TEST(TypeParameters, CreateNumericTypeParametersWithMaxLiteral) {
  NumericTypeParametersProto proto;
  proto.set_is_max_precision(true);
  proto.set_scale(20);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters type_param,
                       TypeParameters::MakeNumericTypeParameters(proto));
  EXPECT_TRUE(type_param.IsNumericTypeParameters());
  EXPECT_EQ(type_param.DebugString(), "(precision=MAX,scale=20)");
}
TEST(TypeParameters, CreateNumericTypeParametersWithPrecisonAndScale) {
  NumericTypeParametersProto proto;
  proto.set_precision(20);
  proto.set_scale(7);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters type_param,
                       TypeParameters::MakeNumericTypeParameters(proto));
  EXPECT_TRUE(type_param.IsNumericTypeParameters());
  EXPECT_EQ(type_param.DebugString(), "(precision=20,scale=7)");
}
TEST(TypeParameters, CreateNumericTypeParametersWithPrecisionOnly) {
  NumericTypeParametersProto proto;
  proto.set_precision(30);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters type_param,
                       TypeParameters::MakeNumericTypeParameters(proto));
  EXPECT_TRUE(type_param.IsNumericTypeParameters());
  EXPECT_EQ(type_param.DebugString(), "(precision=30,scale=0)");
}

TEST(TypeParameters, CreateExtendedTypeParameters) {
  std::vector<SimpleValue> parameters;
  parameters.push_back(SimpleValue::Int64(100));
  parameters.push_back(SimpleValue::String("random"));
  TypeParameters type_param = TypeParameters::MakeExtendedTypeParameters(
      ExtendedTypeParameters(parameters));
  ExtendedTypeParameters extended_type_parameters =
      type_param.extended_type_parameters();

  EXPECT_EQ(extended_type_parameters.num_parameters(), 2);
  EXPECT_EQ(extended_type_parameters.parameter(0), SimpleValue::Int64(100));
  EXPECT_EQ(extended_type_parameters.parameter(1),
            SimpleValue::String("random"));
  EXPECT_EQ(extended_type_parameters.DebugString(), "(100,\"random\")");
}

TEST(TypeParameters, TypeParametersWithChildList) {
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);

  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_child,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  // sub-field without type parameters.
  child_list.push_back(TypeParameters());
  child_list.push_back(string_child);
  child_list.push_back(numeric_child);
  TypeParameters struct_type_param =
      TypeParameters::MakeTypeParametersWithChildList(child_list);
  EXPECT_EQ(struct_type_param.num_children(), 3);
  EXPECT_TRUE(struct_type_param.IsStructOrArrayParameters());
  EXPECT_FALSE(struct_type_param.IsEmpty());
  EXPECT_TRUE(struct_type_param.child(0).IsEmpty());
  EXPECT_EQ(struct_type_param.child(1).string_type_parameters().max_length(),
            10);
  EXPECT_EQ(struct_type_param.child(2).numeric_type_parameters().precision(),
            10);
  EXPECT_EQ(struct_type_param.child(2).numeric_type_parameters().scale(), 5);
  EXPECT_EQ(struct_type_param.DebugString(),
            "[null,(max_length=10),(precision=10,scale=5)]");
}

TEST(TypeParameters, TypeParametersWithSetChildList) {
  TypeParameters params = TypeParameters();

  // Create child_list.
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);

  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_child,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  child_list.push_back(TypeParameters());
  child_list.push_back(string_child);
  child_list.push_back(numeric_child);

  EXPECT_TRUE(params.IsEmpty());
  params.set_child_list(child_list);
  EXPECT_FALSE(params.IsEmpty());
  EXPECT_TRUE(params.IsStructOrArrayParameters());
  EXPECT_EQ(params.num_children(), 3);
  EXPECT_EQ(params.child(1).string_type_parameters().max_length(), 10);
  EXPECT_EQ(params.child(2).numeric_type_parameters().precision(), 10);
  EXPECT_EQ(params.child(2).numeric_type_parameters().scale(), 5);
  EXPECT_EQ(params.DebugString(),
            "[null,(max_length=10),(precision=10,scale=5)]");
}

TEST(TypeParameters, ExtendedTypeParametersWithChildList) {
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  child_list.push_back(string_child);
  std::vector<SimpleValue> parameters;
  parameters.push_back(SimpleValue::Int64(999));
  parameters.push_back(SimpleValue::String("abc"));
  TypeParameters extended_type_with_children =
      TypeParameters::MakeExtendedTypeParameters(
          ExtendedTypeParameters(parameters), child_list);
  EXPECT_TRUE(extended_type_with_children.IsStructOrArrayParameters());
  EXPECT_EQ(extended_type_with_children.num_children(), 1);
  EXPECT_FALSE(extended_type_with_children.IsEmpty());
  EXPECT_EQ(extended_type_with_children.child(0)
                .string_type_parameters()
                .max_length(),
            10);
  EXPECT_EQ(extended_type_with_children.DebugString(),
            "(999,\"abc\")[(max_length=10)]");
}

// Roundtrips TypeParameters through TypeParametersProto and back.
static void SerializeDeserialize(const TypeParameters& type_parameters) {
  TypeParametersProto type_parameters_proto;
  ZETASQL_ASSERT_OK(type_parameters.Serialize(&type_parameters_proto))
      << type_parameters.DebugString();
  auto status_or_type_parameters =
      TypeParameters::Deserialize(type_parameters_proto);
  ZETASQL_ASSERT_OK(status_or_type_parameters.status());
  std::string output;
  google::protobuf::TextFormat::PrintToString(type_parameters_proto, &output);
  EXPECT_TRUE(type_parameters.Equals(status_or_type_parameters.value()))
      << "\nSerialized type_parameters:\n"
      << type_parameters_proto.DebugString();
}

// Roundtrips TypeParametersProto through TypeParameters and back.
static void DeserializeSerialize(const std::string& type_parameters_proto_str) {
  TypeParametersProto type_parameters_proto;
  ZETASQL_CHECK(google::protobuf::TextFormat::ParseFromString(type_parameters_proto_str,
                                             &type_parameters_proto))
      << type_parameters_proto_str;
  auto status_or_value = TypeParameters::Deserialize(type_parameters_proto);
  ZETASQL_ASSERT_OK(status_or_value.status()) << type_parameters_proto.DebugString();
  TypeParametersProto roundtrip_type_parameters_proto;
  ZETASQL_ASSERT_OK(status_or_value.value().Serialize(&roundtrip_type_parameters_proto))
      << roundtrip_type_parameters_proto.DebugString();
  EXPECT_THAT(type_parameters_proto,
              testing::EqualsProto(roundtrip_type_parameters_proto));
}

static void DeserializeWithExpectedError(
    const std::string& type_parameters_proto_str,
    absl::string_view expected_error_message) {
  TypeParametersProto type_parameters_proto;
  ZETASQL_CHECK(google::protobuf::TextFormat::ParseFromString(type_parameters_proto_str,
                                             &type_parameters_proto))
      << type_parameters_proto_str;
  auto status_or_value = TypeParameters::Deserialize(type_parameters_proto);
  EXPECT_EQ(status_or_value.status().code(), absl::StatusCode::kInternal);
  EXPECT_THAT(status_or_value.status().message(),
              ::testing::HasSubstr(expected_error_message));
}

TEST(TypeParameters, SerializeStringTypeParameters) {
  // STRING(L)/STRING(MAX) or BYTES(L)/BYTES(MAX)
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_type_param,
                       TypeParameters::MakeStringTypeParameters(string_param));
  SerializeDeserialize(string_type_param);
  string_param.set_is_max_length(true);
  ZETASQL_ASSERT_OK_AND_ASSIGN(string_type_param,
                       TypeParameters::MakeStringTypeParameters(string_param));
  SerializeDeserialize(string_type_param);
}

TEST(TypeParameters, SerializeNumericTypeParameters) {
  // NUMERIC/BIGNUMERIC(P,S) or NUMERIC/BIGNUMERIC(P) or BIGNUMERIC(MAX) or
  // BIGNUMERIC(MAX,S)
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_type_param,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  SerializeDeserialize(numeric_type_param);
  numeric_param.set_is_max_precision(true);
  numeric_param.set_scale(20);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      numeric_type_param,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  SerializeDeserialize(numeric_type_param);

  // Extended type parameters with EnumValue parameters.
  std::vector<SimpleValue> parameters;
  parameters.push_back(SimpleValue::Int64(100));
  parameters.push_back(SimpleValue::String("random"));
  ExtendedTypeParameters extended_type_param =
      ExtendedTypeParameters(parameters);
  SerializeDeserialize(
      TypeParameters::MakeExtendedTypeParameters(extended_type_param));
}

TEST(TypeParameters, SerializeTypeParametersWithChildList) {
  // Type parameters with child_list.
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto bytes_param;
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(22);
  numeric_param.set_scale(7);
  bytes_param.set_max_length(100);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters bytes_type_param,
                       TypeParameters::MakeStringTypeParameters(bytes_param));
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_type_param,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  // sub-field without type parameters.
  child_list.push_back(TypeParameters());
  child_list.push_back(bytes_type_param);
  child_list.push_back(numeric_type_param);
  TypeParameters struct_type_param =
      TypeParameters::MakeTypeParametersWithChildList(child_list);
  SerializeDeserialize(struct_type_param);
}

TEST(TypeParameters, SerializeExtendedTypeParametersWithChildList) {
  // Type parameters with child_list.
  std::vector<TypeParameters> child_list;
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(22);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_type_param,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  // sub-field without type parameters.
  child_list.push_back(TypeParameters());
  child_list.push_back(numeric_type_param);
  std::vector<SimpleValue> parameters;
  parameters.push_back(SimpleValue::Int64(333));
  parameters.push_back(SimpleValue::Int64(666));
  TypeParameters extended_type_with_children =
      TypeParameters::MakeExtendedTypeParameters(
          ExtendedTypeParameters(parameters), child_list);
  SerializeDeserialize(extended_type_with_children);
}

TEST(TypeParameters, DeserializeStringTypeParametersSuccess) {
  // STRING(L)/STRING(MAX) or BYTES(L)/BYTES(MAX)
  DeserializeSerialize(R"proto(
    string_type_parameters { max_length: 100 })proto");
  DeserializeSerialize(R"proto(
    string_type_parameters { is_max_length: true })proto");
}

TEST(TypeParameters, DeserializeNumericTypeParametersSuccess) {
  // NUMERIC/BIGNUMERIC(P,S) or NUMERIC/BIGNUMERIC(P) or BIGNUMERIC(MAX) or
  // BIGNUMERIC(MAX,S)
  DeserializeSerialize(
      R"proto(
        numeric_type_parameters { precision: 20 scale: 5 })proto");
  DeserializeSerialize(
      R"proto(
        numeric_type_parameters { precision: 60 scale: 30 })proto");
  DeserializeSerialize(
      R"proto(
        numeric_type_parameters { is_max_precision: true scale: 0 })proto");
  DeserializeSerialize(
      R"proto(
        numeric_type_parameters { is_max_precision: true scale: 20 })proto");
}

TEST(TypeParameters, DeserializeExtendedTypeParametersSuccess) {
  // Extended type parameters.
  DeserializeSerialize(
      R"proto(
        extended_type_parameters {
          parameters { int64_value: 100 }
          parameters { string_value: "random" }
        })proto");
}

TEST(TypeParameters, DeserializeTypeParameterWithChildList) {
  // Type parameters with child_list.
  DeserializeSerialize(
      R"proto(
        child_list { string_type_parameters { is_max_length: true } }
        child_list { string_type_parameters { max_length: 300 } }
        child_list { numeric_type_parameters { precision: 20 scale: 5 } })proto");
}

TEST(TypeParameters, DeserializeExtendedTypeParametersWithChildList) {
  // Extended type parameters.
  DeserializeSerialize(
      R"proto(
        extended_type_parameters {
          parameters { int64_value: 100 }
          parameters { string_value: "random" }
        }
        child_list { string_type_parameters { is_max_length: true } }
        child_list { string_type_parameters { max_length: 300 } }
      )proto");
}

TEST(TypeParameters, DeserializeStringTypeParametersFailed) {
  DeserializeWithExpectedError(
      R"proto(
        string_type_parameters { max_length: -100 })proto",
      "max_length must be larger than 0, actual max_length: -100");

  DeserializeWithExpectedError(
      R"proto(
        string_type_parameters { is_max_length: false })proto",
      "is_max_length should either be unset or true");
}

TEST(TypeParameters, DeserializeNumericTypeParametersFailed) {
  DeserializeWithExpectedError(
      R"proto(
        numeric_type_parameters { is_max_precision: false scale: 0 })proto",
      "is_max_precision should either be unset or true");

  DeserializeWithExpectedError(
      R"proto(
        numeric_type_parameters { precision: 100 scale: 30 })proto",
      "precision must be within range [1, 76] or MAX, actual precision: 100");

  DeserializeWithExpectedError(
      R"proto(
        numeric_type_parameters { precision: 50 scale: 40 })proto",
      "scale must be within range [0, 38], actual scale: 40");

  DeserializeWithExpectedError(
      R"proto(
        numeric_type_parameters { precision: 30 scale: 35 })proto",
      "precision must be equal or larger than scale, actual precision: 30, "
      "scale: 35");
}

// StringTypeParameters matches STRING and BYTES.
TEST(TypeParameters, MatchStringOrBytesType) {
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_type_param,
                       TypeParameters::MakeStringTypeParameters(string_param));
  TypeFactory type_factory;
  EXPECT_TRUE(string_type_param.MatchType(type_factory.get_string()));
  EXPECT_TRUE(string_type_param.MatchType(type_factory.get_bytes()));
  EXPECT_FALSE(string_type_param.MatchType(type_factory.get_bool()));
  EXPECT_FALSE(string_type_param.MatchType(type_factory.get_bignumeric()));
}

// NumericTypeParameters matches NUMERIC and BIGNUMERIC.
TEST(TypeParameters, MatchNumericOrBigNumericType) {
  TypeFactory type_factory;
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_type_param,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  EXPECT_TRUE(numeric_type_param.MatchType(type_factory.get_numeric()));
  EXPECT_TRUE(numeric_type_param.MatchType(type_factory.get_bignumeric()));
  EXPECT_FALSE(numeric_type_param.MatchType(type_factory.get_int64()));
}

// Empty TypeParameters matches all types.
TEST(TypeParameters, MatchNonParameterizedType) {
  TypeFactory type_factory;
  const Type* struct_type;
  const Type* array_type;
  ZETASQL_ASSERT_OK(type_factory.MakeStructType({{"int64", type_factory.get_int64()}},
                                        &struct_type));
  ZETASQL_ASSERT_OK(
      type_factory.MakeArrayType(type_factory.get_string(), &array_type));
  TypeParameters empty_type_param = TypeParameters();
  EXPECT_TRUE(empty_type_param.MatchType(type_factory.get_bool()));
  EXPECT_TRUE(empty_type_param.MatchType(type_factory.get_bignumeric()));
  EXPECT_TRUE(empty_type_param.MatchType(struct_type));
  EXPECT_TRUE(empty_type_param.MatchType(array_type));
}

// TypeParameters with child_list matches ARRAY type.
TEST(TypeParameters, MatchTypeParametersWithArrayType) {
  TypeFactory type_factory;

  // Create TypeParameters with child_list with one child.
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  child_list.push_back(string_child);
  TypeParameters array_type_param =
      TypeParameters::MakeTypeParametersWithChildList(child_list);

  // Create array types.
  const Type* string_array = nullptr;
  const Type* numeric_array = nullptr;
  ZETASQL_ASSERT_OK(
      type_factory.MakeArrayType(type_factory.get_string(), &string_array));
  ZETASQL_ASSERT_OK(
      type_factory.MakeArrayType(type_factory.get_numeric(), &numeric_array));

  EXPECT_TRUE(array_type_param.MatchType(string_array));
  EXPECT_FALSE(array_type_param.MatchType(numeric_array));
}

// TypeParameters with a one-child child_list matches STRUCT type.
TEST(TypeParameters, MatchTypeParametersWithOneFieldStructType) {
  TypeFactory type_factory;

  // Create TypeParameters with child_list with one child.
  std::vector<TypeParameters> child_list;
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  child_list.push_back(string_child);
  TypeParameters one_child_type_param =
      TypeParameters::MakeTypeParametersWithChildList(child_list);

  // Create struct/array types.
  const Type* invalid_one_field_struct = nullptr;
  const Type* one_field_struct = nullptr;
  const Type* two_field_struct = nullptr;
  const Type* string_array = nullptr;
  ZETASQL_ASSERT_OK(type_factory.MakeStructType({{"int64", type_factory.get_int64()}},
                                        &invalid_one_field_struct));
  ZETASQL_ASSERT_OK(type_factory.MakeStructType({{"string", type_factory.get_string()}},
                                        &one_field_struct));
  ZETASQL_ASSERT_OK(
      type_factory.MakeStructType({{"string", type_factory.get_string()},
                                   {"numeric", type_factory.get_numeric()}},
                                  &two_field_struct));
  ZETASQL_ASSERT_OK(
      type_factory.MakeArrayType(type_factory.get_string(), &string_array));

  EXPECT_TRUE(one_child_type_param.MatchType(one_field_struct));
  EXPECT_FALSE(one_child_type_param.MatchType(invalid_one_field_struct));
  EXPECT_FALSE(one_child_type_param.MatchType(two_field_struct));
  // A TypeParameters object with one child can be applicable to both an ARRAY
  // and STRUCT type if the STRUCT has one field and that field is the same type
  // as that of the ARRAY elements.
  EXPECT_TRUE(one_child_type_param.MatchType(string_array));
}

// TypeParameters with a two-child child_list matches STRUCT type.
TEST(TypeParameters, MatchTypeParametersWithTwoFieldStructType) {
  TypeFactory type_factory;

  // Create TypeParameters with child_list with two children.
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_child,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  std::vector<TypeParameters> child_list = {string_child, numeric_child};
  TypeParameters two_children_type_param =
      TypeParameters::MakeTypeParametersWithChildList(child_list);

  // Create struct types.
  const Type* one_field_struct = nullptr;
  const Type* invalid_two_field_struct = nullptr;
  const Type* two_field_struct = nullptr;
  ZETASQL_ASSERT_OK(type_factory.MakeStructType({{"string", type_factory.get_string()}},
                                        &one_field_struct));
  ZETASQL_ASSERT_OK(
      type_factory.MakeStructType({{"numeric", type_factory.get_numeric()},
                                   {"string", type_factory.get_string()}},
                                  &invalid_two_field_struct));
  ZETASQL_ASSERT_OK(
      type_factory.MakeStructType({{"string", type_factory.get_string()},
                                   {"numeric", type_factory.get_numeric()}},
                                  &two_field_struct));

  EXPECT_FALSE(two_children_type_param.MatchType(one_field_struct));
  EXPECT_FALSE(two_children_type_param.MatchType(invalid_two_field_struct));
  EXPECT_TRUE(two_children_type_param.MatchType(two_field_struct));
}

// TypeParameters with a nested child_list matches
// STRUCT<ARRAY<STRUCT<STRING, NUMERIC>>, STRING> type.
TEST(TypeParameters, MatchTypeParametersNestedStructType) {
  TypeFactory type_factory;

  // Create nested TypeParameters.
  StringTypeParametersProto string_param;
  string_param.set_max_length(10);
  ZETASQL_ASSERT_OK_AND_ASSIGN(TypeParameters string_child,
                       TypeParameters::MakeStringTypeParameters(string_param));
  NumericTypeParametersProto numeric_param;
  numeric_param.set_precision(10);
  numeric_param.set_scale(5);
  ZETASQL_ASSERT_OK_AND_ASSIGN(
      TypeParameters numeric_child,
      TypeParameters::MakeNumericTypeParameters(numeric_param));
  std::vector<TypeParameters> child_list = {string_child, numeric_child};
  TypeParameters struct_params =
      TypeParameters::MakeTypeParametersWithChildList(child_list);
  child_list = {struct_params};
  TypeParameters array_params =
      TypeParameters::MakeTypeParametersWithChildList(child_list);
  child_list = {array_params, string_child};
  TypeParameters nested_struct_params =
      TypeParameters::MakeTypeParametersWithChildList(child_list);

  // Create nested struct type.
  const Type* child_struct_type = nullptr;
  const Type* child_array_type = nullptr;
  const Type* nested_struct_type = nullptr;
  ZETASQL_ASSERT_OK(
      type_factory.MakeStructType({{"string", type_factory.get_string()},
                                   {"numeric", type_factory.get_numeric()}},
                                  &child_struct_type));
  ZETASQL_ASSERT_OK(
      type_factory.MakeArrayType(child_struct_type, &child_array_type));
  ZETASQL_ASSERT_OK(
      type_factory.MakeStructType({{"array", child_array_type},
                                   {"string", type_factory.get_string()}},
                                  &nested_struct_type));

  EXPECT_TRUE(nested_struct_params.MatchType(nested_struct_type));
  EXPECT_FALSE(nested_struct_params.MatchType(child_struct_type));
}

}  // namespace
}  // namespace zetasql

File: javatests/com/google/zetasql/TypeParametersTest.java
/*
 * Copyright 2019 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.google.zetasql;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.fail;

import com.google.zetasql.ZetaSQLTypeParameters.NumericTypeParametersProto;
import com.google.zetasql.ZetaSQLTypeParameters.StringTypeParametersProto;
import com.google.zetasql.ZetaSQLTypeParameters.TypeParametersProto;

import java.util.Arrays;
import java.util.List;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)

public final class TypeParametersTest {

  @Test
  public void createStringTypeParametersWithMaxLiteral() {
    StringTypeParametersProto proto =
        StringTypeParametersProto.newBuilder().setIsMaxLength(true).build();
    TypeParameters typeParameters = new TypeParameters(proto);
    assertThat(typeParameters.isStringTypeParameters()).isTrue();
    assertThat(typeParameters.getStringTypeParameters().getIsMaxLength()).isTrue();
    assertThat(typeParameters.debugString()).isEqualTo("(max_length=MAX)");
  }

  @Test
  public void createStringTypeParametersWithMaxLength() {
    StringTypeParametersProto proto =
        StringTypeParametersProto.newBuilder().setMaxLength(1000).build();
    TypeParameters typeParameters = new TypeParameters(proto);
    assertThat(typeParameters.isStringTypeParameters()).isTrue();
    assertThat(typeParameters.getStringTypeParameters().getMaxLength()).isEqualTo(1000);
    assertThat(typeParameters.debugString()).isEqualTo("(max_length=1000)");
  }

  @Test
  public void createNumericTypeParametersWithMaxLiteral() {
    NumericTypeParametersProto proto =
        NumericTypeParametersProto.newBuilder().setIsMaxPrecision(true).setScale(20).build();
    TypeParameters typeParameters = new TypeParameters(proto);
    assertThat(typeParameters.isNumericTypeParameters()).isTrue();
    assertThat(typeParameters.debugString()).isEqualTo("(precision=MAX,scale=20)");
  }

  @Test
  public void createNumericTypeParametersWithPrecisonAndScale() {
    NumericTypeParametersProto proto =
        NumericTypeParametersProto.newBuilder().setPrecision(20).setScale(7).build();
    TypeParameters typeParameters = new TypeParameters(proto);
    assertThat(typeParameters.isNumericTypeParameters()).isTrue();
    assertThat(typeParameters.debugString()).isEqualTo("(precision=20,scale=7)");
  }

  @Test
  public void createNumericTypeParametersWithPrecisionOnly() {
    NumericTypeParametersProto proto =
        NumericTypeParametersProto.newBuilder().setPrecision(20).build();
    TypeParameters typeParameters = new TypeParameters(proto);
    assertThat(typeParameters.isNumericTypeParameters()).isTrue();
    assertThat(typeParameters.debugString()).isEqualTo("(precision=20,scale=0)");
  }

  @Test
  public void createTypeParametersWithChildList() {
    StringTypeParametersProto stringParam =
        StringTypeParametersProto.newBuilder().setMaxLength(10).build();
    NumericTypeParametersProto numericParam =
        NumericTypeParametersProto.newBuilder().setPrecision(10).setScale(5).build();
    List<TypeParameters> childList =
        Arrays.asList(
            new TypeParameters(),
            new TypeParameters(stringParam),
            new TypeParameters(numericParam));
    TypeParameters structParam = new TypeParameters(childList);
    assertThat(structParam.getChildCount()).isEqualTo(3);
    assertThat(structParam.isStructOrArrayParameters()).isTrue();
    assertThat(structParam.isEmpty()).isFalse();
    assertThat(structParam.debugString())
        .isEqualTo("[null,(max_length=10),(precision=10,scale=5)]");
  }

  public static void checkSerializeAndDeserialize(TypeParameters typeParameters) {
    try {
      TypeParameters typeParameters2 = TypeParameters.deserialize(typeParameters.serialize());
      assertThat(typeParameters).isEqualTo(typeParameters2);
      assertThat(typeParameters2).isEqualTo(typeParameters);
      assertThat(typeParameters.hashCode()).isEqualTo(typeParameters2.hashCode());
      assertThat(typeParameters.serialize()).isEqualTo(typeParameters2.serialize());
    } catch (IllegalArgumentException e) {
      fail(e.toString());
    }
  }

  @Test
  public void serializeAndDeserializeStringTypeParameters() {
    checkSerializeAndDeserialize(
        new TypeParameters(StringTypeParametersProto.newBuilder().setIsMaxLength(true).build()));
    checkSerializeAndDeserialize(
        new TypeParameters(StringTypeParametersProto.newBuilder().setMaxLength(10).build()));
  }

  @Test
  public void serializeAndDeserializeNumericTypeParameters() {
    checkSerializeAndDeserialize(
        new TypeParameters(
            NumericTypeParametersProto.newBuilder().setPrecision(10).setScale(5).build()));
    checkSerializeAndDeserialize(
        new TypeParameters(
            NumericTypeParametersProto.newBuilder().setIsMaxPrecision(true).setScale(20).build()));
  }

  @Test
  public void serializeAndDeserializeTypeParametersWithChildList() {
    StringTypeParametersProto stringParam =
        StringTypeParametersProto.newBuilder().setMaxLength(50).build();
    NumericTypeParametersProto numericParam =
        NumericTypeParametersProto.newBuilder().setPrecision(70).setScale(30).build();
    List<TypeParameters> childList =
        Arrays.asList(new TypeParameters(stringParam), new TypeParameters(numericParam));
    TypeParameters structParam = new TypeParameters(childList);
    checkSerializeAndDeserialize(structParam);
  }

  @Test
  public void deserializeStringTypeParametersFailed() {
    // is_max_length is set to false.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setStringTypeParameters(StringTypeParametersProto.newBuilder().setIsMaxLength(false))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected)
          .hasMessageThat()
          .contains("is_max_length should either be unset or true");
    }

    // max_length <= 0.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setStringTypeParameters(StringTypeParametersProto.newBuilder().setMaxLength(-100))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected).hasMessageThat().contains("max_length must be larger than 0");
    }
  }

  @Test
  public void deserializeNumericTypeParametersFailed() {
    // is_max_precision is set to false.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setNumericTypeParameters(
                  NumericTypeParametersProto.newBuilder().setIsMaxPrecision(false))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected)
          .hasMessageThat()
          .contains("is_max_precision should either be unset or true");
    }

    // precision > 76.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setNumericTypeParameters(
                  NumericTypeParametersProto.newBuilder().setPrecision(100).setScale(30))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected)
          .hasMessageThat()
          .contains("precision must be within range [1, 76] or MAX");
    }

    // scale > 38.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setNumericTypeParameters(
                  NumericTypeParametersProto.newBuilder().setPrecision(50).setScale(40))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected).hasMessageThat().contains("scale must be within range [0, 38]");
    }

    // precision < scale.
    try {
      TypeParameters.deserialize(
          TypeParametersProto.newBuilder()
              .setNumericTypeParameters(
                  NumericTypeParametersProto.newBuilder().setPrecision(30).setScale(35))
              .build());
      fail();
    } catch (IllegalArgumentException expected) {
      assertThat(expected)
          .hasMessageThat()
          .contains("precision must be equal to or larger than scale");
    }
  }

  @Test
  public void twoTypeParametersAreEqual() {
    TypeParameters numericParam1 =
        new TypeParameters(
            NumericTypeParametersProto.newBuilder().setIsMaxPrecision(true).setScale(20).build());
    TypeParameters numericParam2 =
        new TypeParameters(
            NumericTypeParametersProto.newBuilder().setIsMaxPrecision(true).setScale(20).build());
    TypeParameters numericParam3 =
        new TypeParameters(
            NumericTypeParametersProto.newBuilder().setPrecision(50).setScale(20).build());
    assertThat(numericParam1.equals(numericParam2)).isTrue();
    assertThat(numericParam1.equals(numericParam3)).isFalse();

    TypeParameters stringParam1 =
        new TypeParameters(StringTypeParametersProto.newBuilder().setMaxLength(100).build());
    TypeParameters stringParam2 =
        new TypeParameters(StringTypeParametersProto.newBuilder().setMaxLength(100).build());
    TypeParameters stringParam3 =
        new TypeParameters(StringTypeParametersProto.newBuilder().setMaxLength(1000).build());
    assertThat(stringParam1.equals(stringParam2)).isTrue();
    assertThat(stringParam1.equals(stringParam3)).isFalse();

    List<TypeParameters> childList1 =
        Arrays.asList(stringParam1, new TypeParameters(), numericParam1);
    List<TypeParameters> childList2 =
        Arrays.asList(stringParam2, new TypeParameters(), numericParam2);
    List<TypeParameters> childList3 = Arrays.asList(stringParam3, numericParam2);
    assertThat(new TypeParameters(childList1).equals(new TypeParameters(childList2))).isTrue();
    assertThat(new TypeParameters(childList1).equals(new TypeParameters(childList3))).isFalse();
  }
}
